### PR TITLE
Preprocessor optimizations

### PR DESCRIPTION
**Pull request by guykatzz**
Merge two consecutive weighted sum layers into one

### CODE REVIEW

File: src/engine/Engine.cpp [hunk 1]
```diff
@@ -1053,7 +1053,9 @@ void Engine::initializeNetworkLevelReasoning()
     _networkLevelReasoner = _preprocessedQuery.getNetworkLevelReasoner();
 
     if ( _networkLevelReasoner )
-        _networkLevelReasoner->setTableau( _tableau );
```
Comment (guykatzz) [inline src/engine/Engine.cpp:1056]: Please undo this change

Comment (guyam2) [reply src/engine/Engine.cpp:1056]: OK. And commit + push after that?

Comment (guykatzz) [reply src/engine/Engine.cpp:1056]: Nope, still here... You can ad should look at the "Files changes" tab to see the actual changes

File: src/engine/InputQuery.cpp [hunk 2]
```diff
@@ -590,15 +590,24 @@ bool InputQuery::constructNetworkLevelReasoner()
     List<unsigned> inputs = getInputVariables();
     nlr->addLayer( 0, NLR::Layer::INPUT, inputs.size() );
     unsigned index = 0;
+
+    auto *layer = nlr->getLayerToUpdate( 0 );
```
Comment (guykatzz) [inline src/engine/InputQuery.cpp:594]: Change from "auto *layer" to Layer *inputLayer

We try to use auto only in for loops
And informative names are better

File: src/engine/Preprocessor.cpp [hunk 3]
```diff
@@ -44,10 +44,24 @@ InputQuery Preprocessor::preprocess( const InputQuery &query, bool attemptVariab
     makeAllEquationsEqualities();
 
     /*
-      Attempt to construct a network level reasonor
+      Attempt to construct a network level reasoner
     */
     _preprocessed.constructNetworkLevelReasoner();
 
+    /*
+      Merge subsequent WS layers
```
Comment (guykatzz) [inline src/engine/Preprocessor.cpp:52]: change subsequent to consecutive in the comment. It emphasizes the fact that they are immediately adjacent

File: src/engine/Preprocessor.cpp [hunk 4]
```diff
@@ -560,14 +576,24 @@ void Preprocessor::collectFixedValues()
...
+            _preprocessed.setLowerBound( i, _fixedVariables[i] );
+            _preprocessed.setUpperBound( i, _fixedVariables[i] );
         }
-	}
+    }
 }
 
 void Preprocessor::eliminateVariables()
 {
+    printf( "PP: eliminateVariables starting!\n" );
```
Comment (guykatzz) [inline src/engine/Preprocessor.cpp:595]: Undo change - printfs shouldn't be merged to master, debug printing should be made using the LOG macros that can be turned on/off

File: src/nlr/CMakeLists.txt [hunk 5]
```diff
@@ -15,6 +15,9 @@ macro(network_level_reasoner_add_unit_test name)
 endmacro()
 
 network_level_reasoner_add_unit_test(NetworkLevelReasoner)
+network_level_reasoner_add_unit_test(WsLayerElimination)
+
```
Comment (guykatzz) [inline src/nlr/CMakeLists.txt:19]: Remove the empty lines

File: src/nlr/Layer.cpp [hunk 6]
```diff
@@ -1481,4 +1501,142 @@ double Layer::getEliminatedNeuronValue( unsigned neuron ) const
...
+
+void Layer::adjustWeightMapIndexing( Map<unsigned, double *> &map, unsigned startIndex )
+{
+    Map<unsigned, double *> copyOfWeights = map;
+    map.clear();
+    for ( const auto &pair : copyOfWeights )
+        map[pair.first >= startIndex ? pair.first - 1 : pair.first] = pair.second;
+}
+
+void Layer::reduceIndexAfterMerge (unsigned startIndex)
```
Comment (guykatzz) [inline src/nlr/Layer.cpp:1536]: Spacing:

void Layer::reduceIndexAfterMerge (unsigned startIndex)
-->
void Layer::reduceIndexAfterMerge( unsigned startIndex )

File: src/nlr/Layer.cpp [hunk 7]
```diff
@@ -1481,4 +1501,142 @@ double Layer::getEliminatedNeuronValue( unsigned neuron ) const
...
+{
+    if ( _layerIndex >= startIndex )
+    {
+        _layerIndex--;
```
Comment (guykatzz) [inline src/nlr/Layer.cpp:1540]: Always prefer --i to i--. I can explain why in person.
Also, no need to use { and } for such simple cases.

File: src/nlr/Layer.cpp [hunk 8]
```diff
@@ -1481,4 +1501,142 @@ double Layer::getEliminatedNeuronValue( unsigned neuron ) const
...
+    }
+}
+
+bool Layer::operator==( const Layer & layer ) const
+{
+    if ( _layerIndex != layer._layerIndex )
+    {
```
Comment (guykatzz) [inline src/nlr/Layer.cpp:1547]: in all these if statements, remove { and }

File: src/nlr/Layer.cpp [hunk 9]
```diff
@@ -1481,4 +1501,142 @@ double Layer::getEliminatedNeuronValue( unsigned neuron ) const
...
+        return false;
+    }
+
+    if ( ( _bias && !layer._bias ) || ( !_bias && layer._bias ) )
+    {
+        return false;
+    }
+
+    if (_bias && layer._bias )
```
Comment (guykatzz) [inline src/nlr/Layer.cpp:1571]: (_bias --> ( _bias

File: src/nlr/Layer.cpp [hunk 10]
```diff
@@ -1481,4 +1501,142 @@ double Layer::getEliminatedNeuronValue( unsigned neuron ) const
...
+    {
+        if ( std::memcmp(_bias, layer._bias, _size ) != 0 )
```
Comment (guykatzz) [inline src/nlr/Layer.cpp:1573]: memcmp(_bias --> memcmp( _bias

File: src/nlr/Layer.cpp [hunk 11]
```diff
@@ -1481,4 +1501,142 @@ double Layer::getEliminatedNeuronValue( unsigned neuron ) const
...
+            return false;
+        }
+    }
+
+    if ( !compareWights( _layerToWeights, layer._layerToWeights ) )
+    {
+        return false;
+    }
+
+    if (!compareWights( _layerToPositiveWeights, layer._layerToPositiveWeights ) )
```
Comment (guykatzz) [inline src/nlr/Layer.cpp:1584]: add space before !

File: src/nlr/Layer.cpp [hunk 12]
```diff
@@ -1481,4 +1501,142 @@ double Layer::getEliminatedNeuronValue( unsigned neuron ) const
...
+        return false;
+    }
+
+    return true;
+}
+
+bool Layer::compareWights( const Map<unsigned, double*> &map, const Map<unsigned, double*> &mapOfOtherLayer ) const
+{
+    if ( map.size() != mapOfOtherLayer.size() )
+    {
```
Comment (guykatzz) [inline src/nlr/Layer.cpp:1605]: remove braces

File: src/nlr/Layer.cpp [hunk 13]
```diff
@@ -1481,4 +1501,142 @@ double Layer::getEliminatedNeuronValue( unsigned neuron ) const
...
+
+    if ( _sourceLayers != layer._sourceLayers )
+    {
+        return false;
+    }
+
+    return true;
+}
+
+bool Layer::compareWights( const Map<unsigned, double*> &map, const Map<unsigned, double*> &mapOfOtherLayer ) const
```
Comment (guykatzz) [inline src/nlr/Layer.cpp:1602]: double* --> double *

File: src/nlr/Layer.cpp [hunk 14]
```diff
@@ -1481,4 +1501,142 @@ double Layer::getEliminatedNeuronValue( unsigned neuron ) const
...
+{
+    if ( map.size() != mapOfOtherLayer.size() )
+    {
+        return false;
+    }
+
+    for ( const auto &pair : map )
+    {
+        unsigned key = pair.first;
+        double* value = pair.second;
```
Comment (guykatzz) [inline src/nlr/Layer.cpp:1612]: double* value --> double *value

File: src/nlr/Layer.cpp [hunk 15]
```diff
@@ -1481,4 +1501,142 @@ double Layer::getEliminatedNeuronValue( unsigned neuron ) const
...
+
+        if ( !mapOfOtherLayer.exists( key ) )
+        {
+            return false;
+        }
+
+        if ( sizeof( *value ) / sizeof( double ) != sizeof( *mapOfOtherLayer[key] ) / sizeof( double ) )
```
Comment (guykatzz) [inline src/nlr/Layer.cpp:1619]: Not sure what you were trying to check. I think sizeof( *value ) will just be sizeof( double ), so always 1.

File: src/nlr/Layer.cpp [hunk 16]
```diff
@@ -1481,4 +1501,142 @@ double Layer::getEliminatedNeuronValue( unsigned neuron ) const
...
+        {
+            return false;
+        }
+
+        if ( std::memcmp( value, mapOfOtherLayer[key], sizeof( *value ) / sizeof( double ) ) != 0 )
```
Comment (guykatzz) [inline src/nlr/Layer.cpp:1624]: Same comment, sizeof( *value ) doesn't give you the size of the array

File: src/nlr/Layer.h [hunk 17]
```diff
@@ -53,8 +53,10 @@ class Layer
 
     void setLayerOwner( LayerOwner *layerOwner );
     void addSourceLayer( unsigned layerNumber, unsigned layerSize );
+    void removeSourceLayer( unsigned sourceLayer );
     const Map<unsigned, unsigned> &getSourceLayers() const;
-
+    const double *getWeightMatrix( unsigned sourceLayer ) const;
+    void reduceIndexFromAllMaps( unsigned startIndex );
```
Comment (guykatzz) [inline src/nlr/Layer.h:59]: reduceIndexFromAllMaps is not an intuitive function, maybe add a comment about what it does

File: src/nlr/Layer.h [hunk 18]
```diff
@@ -107,15 +110,18 @@ class Layer
...
+    void reduceIndexAfterMerge( unsigned startIndex );
 
     /*
       For debugging purposes
     */
     void dump() const;
     static String typeToString( Type type );
+    bool operator==( const Layer & layer ) const;
+    bool compareWights( const Map<unsigned, double*> &map, const Map<unsigned, double*> &mapOfOtherLayer ) const;
+
```
Comment (guykatzz) [inline src/nlr/Layer.h:124]: remove extra empty line

File: src/nlr/Layer.h [hunk 19]
```diff
@@ -107,15 +110,18 @@ class Layer
...
     bool neuronEliminated( unsigned neuron ) const;
     double getEliminatedNeuronValue( unsigned neuron ) const;
+    void reduceIndexAfterMerge( unsigned startIndex );
 
     /*
       For debugging purposes
     */
     void dump() const;
     static String typeToString( Type type );
+    bool operator==( const Layer & layer ) const;
```
Comment (guykatzz) [inline src/nlr/Layer.h:122]: Layer & layer --> Layer &layer

File: src/nlr/NetworkLevelReasoner.cpp [hunk 20]
```diff
@@ -77,17 +82,22 @@ void NetworkLevelReasoner::setBias( unsigned layer, unsigned neuron, double bias
...
-    _layerIndexToLayer[targetLeyer]->addActivationSource( sourceLayer, sourceNeuron, targetNeuron );
+    _layerIndexToLayer[targetLayer]->addActivationSource( sourceLayer, sourceNeuron, targetNeuron );
 }
 
 const Layer *NetworkLevelReasoner::getLayer( unsigned index ) const
 {
     return _layerIndexToLayer[index];
 }
 
+Layer *NetworkLevelReasoner::getLayerToUpdate( unsigned index )
```
Comment (guykatzz) [inline src/nlr/NetworkLevelReasoner.cpp:96]: Maybe just call this function "getLayer"?

File: src/nlr/NetworkLevelReasoner.cpp [hunk 21]
```diff
@@ -358,4 +395,152 @@ void NetworkLevelReasoner::generateInputQueryForWeightedSumLayer( InputQuery &in
...
+
+        unsigned inputDimension = inputLayerToFirst->getSize();
+        unsigned middleDimension = firstLayer->getSize();
+        unsigned outputDimension = secondLayer->getSize();
+
+        // Compute new weights
+        const double *firstLayerMatrix = firstLayer->getWeightMatrix( previousToFirstLayerIndex );
+        const double *secondLayerMatrix = secondLayer->getWeightMatrix( firstLayerIndex );
+
+        double *newWeightMatrix = multiplyWeights(firstLayerMatrix,
```
Comment (guykatzz) [inline src/nlr/NetworkLevelReasoner.cpp:470]: space after ( and before )

File: src/nlr/NetworkLevelReasoner.cpp [hunk 22]
```diff
@@ -358,4 +395,152 @@ void NetworkLevelReasoner::generateInputQueryForWeightedSumLayer( InputQuery &in
...
+        }
+
+        // Adjust the sources of the second layer
+        secondLayer->addSourceLayer( previousToFirstLayerIndex, inputLayerToFirst->getSize() );
+        for ( unsigned sourceNeuron = 0; sourceNeuron < inputDimension; ++sourceNeuron )
+        {
+            for ( unsigned targetNeuron = 0; targetNeuron < outputDimension; ++targetNeuron )
+            {
+                double weight = newWeightMatrix[sourceNeuron * outputDimension + targetNeuron];
+                secondLayer->setWeight(previousToFirstLayerIndex, sourceNeuron, targetNeuron,
```
Comment (guykatzz) [inline src/nlr/NetworkLevelReasoner.cpp:494]: space after ( and before )

File: src/nlr/NetworkLevelReasoner.cpp [hunk 23]
```diff
@@ -358,4 +395,152 @@ void NetworkLevelReasoner::generateInputQueryForWeightedSumLayer( InputQuery &in
...
+                                       weight);
+            }
+        }
+
+        delete[] newWeightMatrix; // was without []
```
Comment (guykatzz) [inline src/nlr/NetworkLevelReasoner.cpp:499]: remove comment

File: src/nlr/NetworkLevelReasoner.cpp [hunk 24]
```diff
@@ -358,4 +395,152 @@ void NetworkLevelReasoner::generateInputQueryForWeightedSumLayer( InputQuery &in
...
+    // Remove the first layer from second layer's sources
+    secondLayer->removeSourceLayer( firstLayerIndex );
+
+    // Finally, remove the first layer from the map and delete it
+    _layerIndexToLayer.erase( firstLayerIndex );
+    delete firstLayer;
+
+    // Adjust the indices of all layers starting from secondLayerIndex
+    // and higher
+    for (unsigned i = secondLayerIndex; i <= lastLayerIndex; ++i)
```
Comment (guykatzz) [inline src/nlr/NetworkLevelReasoner.cpp:511]: spaces after ( and before ), remove curly braces

File: src/nlr/NetworkLevelReasoner.cpp [hunk 25]
```diff
@@ -358,4 +395,152 @@ void NetworkLevelReasoner::generateInputQueryForWeightedSumLayer( InputQuery &in
...
+{
+    // update Layer-level maps
+    _layerIndexToLayer[layer]->reduceIndexFromAllMaps( startIndex );
+    _layerIndexToLayer[layer]->reduceIndexAfterMerge( startIndex );
+
+    // Update the mapping in the NLR
+    _layerIndexToLayer[layer - 1] = _layerIndexToLayer[layer];
+    _layerIndexToLayer.erase( layer );
+}
+
```
Comment (guykatzz) [inline src/nlr/NetworkLevelReasoner.cpp:544]: remove extra empty line

File: src/nlr/NetworkLevelReasoner.h [hunk 26]
```diff
@@ -57,6 +58,7 @@ class NetworkLevelReasoner : public LayerOwner
 
     unsigned getNumberOfLayers() const;
     const Layer *getLayer( unsigned index ) const;
+    Layer * getLayerToUpdate( unsigned index );
```
Comment (guykatzz) [inline src/nlr/NetworkLevelReasoner.h:61]: remove space after *

File: src/nlr/tests/Test_WsLayerElimination.h [hunk 27]
```diff
@@ -0,0 +1,839 @@
+
+#ifndef MARABOU_TEST_WSLAYERELIMINATION_H
```
Comment (guykatzz) [inline src/nlr/tests/Test_WsLayerElimination.h:2]: Don't need this in test files

File: src/nlr/tests/Test_WsLayerElimination.h [hunk 28]
```diff
@@ -0,0 +1,839 @@
...
+ ** All rights reserved. See the file COPYING in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** [[ Add lengthier description here ]]
+
+**/
+
+#include <cxxtest/TestSuite.h>
+
+#include "../../engine/tests/MockTableau.h" // TODO: fix this
```
Comment (guykatzz) [inline src/nlr/tests/Test_WsLayerElimination.h:23]: Do the todo :)

File: src/nlr/tests/Test_WsLayerElimination.h [hunk 29]
```diff
@@ -0,0 +1,839 @@
...
+#include "FloatUtils.h"
+#include "InputQuery.h"
+#include "Layer.h"
+#include "NetworkLevelReasoner.h"
+#include "Tightening.h"
+#include "Map.h"
+
+
```
Comment (guykatzz) [inline src/nlr/tests/Test_WsLayerElimination.h:31]: remove extra empty lines, remove comment

File: src/nlr/tests/Test_WsLayerElimination.h [hunk 30]
```diff
@@ -0,0 +1,839 @@
...
+        {
+            TS_ASSERT( *nlr.getLayer( layerNumber ) == *expectedNlr.getLayer( layerNumber ) )
+        }
+
+        // Check the NN outputs are correct
+        double *input = new double[2];
+        double *output = new double[2];
+        double *expectedOutput = new double[2];
+
+        for (auto i = -250; i < 250; ++i)
```
Comment (guykatzz) [inline src/nlr/tests/Test_WsLayerElimination.h:151]: spaces after ( and before ). Use int instead of auto. A few lines down, space before and after "/"

File: src/nlr/tests/Test_WsLayerElimination.h [hunk 31]
```diff
@@ -0,0 +1,839 @@
...
+
+            TS_ASSERT_EQUALS( std::memcmp( output, expectedOutput, 2 ), 0 )
+        }
+
+        delete[] input;
+        delete[] output;
+        delete[] expectedOutput;
+    }
+
+    // Test number 4
```
Comment (guykatzz) [inline src/nlr/tests/Test_WsLayerElimination.h:623]: remove comment

File: src/nlr/tests/Test_WsLayerElimination.h [hunk 32]
```diff
@@ -0,0 +1,839 @@
...
+            TS_ASSERT_EQUALS( std::memcmp( output, expectedOutput, 2 ), 0 )
+        }
+
+        delete[] input;
+        delete[] output;
+        delete[] expectedOutput;
+    }
+
+};
+
```
Comment (guykatzz) [inline src/nlr/tests/Test_WsLayerElimination.h:836]: remove extra empty lines

File: src/nlr/tests/Test_WsLayerElimination.h [hunk 33]
```diff
@@ -0,0 +1,839 @@
...
+        NLR::NetworkLevelReasoner expectedNlr;
+        populateNetworkAfterMerge_CaseD( expectedNlr );
+
+        // Check the NN outputs are correct
+        double *input = new double[2];
+        double *output = new double[3];
+        double *expectedOutput = new double[3];
+
+        for ( auto i = -250; i < 250; ++i ) // was -25 -> 25
```
Comment (guykatzz) [inline src/nlr/tests/Test_WsLayerElimination.h:819]: remove comment

File: src/system_tests/Test_wsElimination.h [hunk 34]
```diff
@@ -0,0 +1,189 @@
...
+ ** in the top-level source directory) and their institutional affiliations.
+ ** All rights reserved. See the file COPYING in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** [[ Add lengthier description here ]]
+
+**/
+
+
+#ifndef MARABOU_TEST_WSELIMINATION_H
```
Comment (guykatzz) [inline src/system_tests/Test_wsElimination.h:17]: don't need in test files, also remove extra empty lines

File: src/system_tests/Test_wsElimination.h [hunk 35]
```diff
@@ -0,0 +1,189 @@
...
+
+#include <cxxtest/TestSuite.h>
+#include "AcasParser.h"
+#include "Engine.h"
+#include "FloatUtils.h"
+#include "GlobalConfiguration.h"
+#include "InputQuery.h"
+#include "Preprocessor.h"
+#include "QueryLoader.h"
+#include "MarabouError.h"
```
Comment (guykatzz) [inline src/system_tests/Test_wsElimination.h:30]: We sort the includes alphabetically. MarabouError is out of place

File: src/system_tests/Test_wsElimination.h [hunk 36]
```diff
@@ -0,0 +1,189 @@
...
+        TS_ASSERT_EQUALS( inputQuery.getSolutionValue( 786 ), 10 );
+        TS_ASSERT_EQUALS( inputQuery.getSolutionValue( 787 ), 16 );
+        TS_ASSERT_EQUALS( inputQuery.getSolutionValue( 788 ), -18 );
+        TS_ASSERT_EQUALS( inputQuery.getSolutionValue( 789 ), 2 );
+        TS_ASSERT_EQUALS( inputQuery.getSolutionValue( 790 ), 6 );
+        TS_ASSERT_EQUALS( inputQuery.getSolutionValue( 791 ), 0 );
+        TS_ASSERT_EQUALS( inputQuery.getSolutionValue( 792 ), 2 );
+        TS_ASSERT_EQUALS( inputQuery.getSolutionValue( 793 ), 2 );
+    }
+
```
Comment (guykatzz) [inline src/system_tests/Test_wsElimination.h:76]: Remove extra space

File: src/engine/Engine.cpp (hunk 1 above)
Comment (guyam2) [reply src/engine/Engine.cpp:1056]: I think I didn't understand - you want me to remove the **whole** lines (if/else) or just the topology dump? I thought you want to revert it back and remove only the dump

File: src/system_tests/Test_wsElimination.h (hunk 35 above)
Comment (guyam2) [reply src/system_tests/Test_wsElimination.h:30]: OK, I originally copied from the acas system test so I fixed it in both mine and the acas test now

File: src/nlr/NetworkLevelReasoner.cpp (hunk 20 above)
Comment (guyam2) [reply src/nlr/NetworkLevelReasoner.cpp:96]: It causes problems because there is a **const** getLayer(). Or do you mean to overload with a non-const version?

File: src/nlr/tests/Test_WsLayerElimination.h (hunk 28 above)
Comment (guyam2) [reply src/nlr/tests/Test_WsLayerElimination.h:23]: copied from another test:) It's fine in ours (I think)

File: src/nlr/Layer.cpp (hunk 15 above)
Comment (guyam2) [reply src/nlr/Layer.cpp:1619]: Actually it's the number of bytes in the array. And in order to have the number of elements I need to divide it byte the size of a single double

Comment (guyam2) [reply src/nlr/Layer.cpp:1619]: I need the number of elements in the array. so you suggest changing it?

Comment (guyam2) [reply src/nlr/Layer.cpp:1619]: OK I thought about it and we can remove these lines completely. I'll explain:
The size of each double array is the size of the current layer multiplied by the size of the input layer. BUT - we already compared (a few lines before) the sizes of the layers, and once we compare keys (input layer indices before these lines) we know the same input layer.

**Review by guyam2 — APPROVED** [review]
